### PR TITLE
Faster Jacobian via Split backends

### PR DESCRIPTION
**Pull request by avik-pal**
## Goals

- [x] ~Fast Code Path for TwoPointBVProblem (Banded Matrix so we can do complete sparse diff)~
  - [x] ~Rewrite the loss function to reorder the outputs correctly.~
- [x] Split Mode
  - [x] Sparse Diff for the Banded Part
  - [x] Dense Diff for the `bc!` Part
    * We won't be able to use Zygote since `bc!` assumes in-place operations
    * The Jacobian has `inputs > outputs` so we want to use ReverseMode ideally. `Enzyme` might be a good choice. For fallback, we can just use `ReverseDiff`.
- [x] Sparsity Detection for `bc!` and if sparse, then use SparseDiff here as well
  - [x] Requires Sparse ReverseMode (Not implemented upstream yet)

### CODE REVIEW

**Comment by avik-pal** [comment]
An initial version of Sparse Reverse Mode AD with Zygote:

```julia
using Pkg
Pkg.activate(@__DIR__)

using BenchmarkTools, BoundaryValueDiffEq, BVProblemLibrary
using Symbolics, ArrayInterface
using FiniteDiff, ForwardDiff, SparseDiffTools
using Zygote
using LinearAlgebra
using Test

function jacobian_sparsity(f, x)
    f!(y, x) = (y .= f(x))
    return jacobian_sparsity(f!, f(x), x)
end

jacobian_sparsity(f!, y, x) = Symbolics.jacobian_sparsity(f!, y, x)

f1(x) = [x[1], x[2] * x[3], x[3]]

x_in = [1.0, 3.0, 2.0]

module ABC
    struct ReverseModeJacobianCache{S, C, I, NR, NC}
        sparsity::S
        colorvec::C
        idx_vec::I
        nz_rows::NR
        nz_cols::NC
    end
end

function sparse_jacobian_setup(f, x)
    y = f(x)
    Jₛ = jacobian_sparsity(f, x)
    (nz_rows, nz_cols) = collect.(ArrayInterface.findstructralnz(Jₛ))
    colorvec = matrix_colors(Jₛ')
    return ABC.ReverseModeJacobianCache(Jₛ, colorvec, collect(1:size(Jₛ, 1)), nz_rows,
        nz_cols)
end

function sparse_jacobian_reversemode(f, x)
    cache = sparse_jacobian_setup(f, x)
    J = similar(cache.sparsity, eltype(x))
    return sparse_jacobian_reversemode!(J, f, x, cache)
end

function sparse_jacobian_reversemode!(J, f, x, cache::ABC.ReverseModeJacobianCache)
    partial_f(x, idx_vec) = dot(f(x), idx_vec)
    for c in 1:maximum(cache.colorvec)
        @. cache.idx_vec = cache.colorvec == c
        gs = first(Zygote.gradient(partial_f, x, cache.idx_vec))
        pick_idxs = [i for i in 1:length(cache.nz_rows) if cache.colorvec[cache.nz_rows[i]] == c]
        row_idxs = cache.nz_rows[pick_idxs]
        col_idxs = cache.nz_cols[pick_idxs]
        len_cols = length(col_idxs)
        unused_cols = setdiff(1:size(J, 2), col_idxs)
        perm_cols = sortperm(vcat(col_idxs, unused_cols))
        row_idxs = vcat(row_idxs, zeros(Int, size(J, 2) - len_cols))[perm_cols]
        for i in axes(J, 1), j in axes(J, 2)
            i == row_idxs[j] && (J[i, j] = gs[j])
        end
    end
    return J
end

cache = sparse_jacobian_setup(f1, x_in)
J = zeros(3, 3)
sparse_jacobian_reversemode!(J, f1, x_in, cache)

sparse_jacobian_reversemode(f1, x_in)

@benchmark Zygote.jacobian($f1, $x_in)

@test Array(sparse_jacobian_reversemode(f1, x_in)) ≈ first(Zygote.jacobian(f1, x_in))
@benchmark sparse_jacobian_reversemode($f1, $x_in)

@test sparse_jacobian_reversemode!(J, f1, x_in, cache) ≈ first(Zygote.jacobian(f1, x_in))
@benchmark sparse_jacobian_reversemode!($J, $f1, $x_in, $cache)
```

Results:

```julia
julia> @benchmark Zygote.jacobian($f1, $x_in)
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  21.010 μs … 202.265 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     27.285 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   26.745 μs ±   5.267 μs  ┊ GC (mean ± σ):  0.00% ± 0.00%

    ▄█▇▄▁                ▁▁▁                                    
  ▂▇█████▆▅▄▃▂▂▂▂▂▂▂▃▄▅▇██████▇▆▅▄▄▃▃▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▁▂▁▁▁▁▁ ▃
  21 μs           Histogram: frequency by time         38.6 μs <

 Memory estimate: 6.03 KiB, allocs estimate: 177.

julia> @test Array(sparse_jacobian_reversemode(f1, x_in)) ≈ first(Zygote.jacobian(f1, x_in))
Test Passed

julia> @benchmark sparse_jacobian_reversemode($f1, $x_in)
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  35.469 μs … 226.444 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     42.254 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   46.914 μs ±  11.458 μs  ┊ GC (mean ± σ):  0.00% ± 0.00%

    ▄▅▇█▆▅▂                                                     
  ▂▆████████▆▅▄▄▃▃▃▃▃▄▄▅▄▃▃▃▃▃▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▁▁▁▁▁▁▁▁▁▁▁ ▃
  35.5 μs         Histogram: frequency by time         84.5 μs <

 Memory estimate: 23.52 KiB, allocs estimate: 363.

julia> @test sparse_jacobian_reversemode!(J, f1, x_in, cache) ≈ first(Zygote.jacobian(f1, x_in))
Test Passed

julia> @benchmark sparse_jacobian_reversemode!($J, $f1, $x_in, $cache)
BenchmarkTools.Trial: 10000 samples with 7 evaluations.
 Range (min … max):  3.924 μs … 17.234 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     5.860 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   6.307 μs ±  1.623 μs  ┊ GC (mean ± σ):  0.00% ± 0.00%

        ▄██▄▃▄▆▅▄▃▁                                           
  ▂▅▃▂▂▆███████████▇▇▆▆▆▆▅▄▄▄▄▄▃▃▃▃▃▃▂▂▂▂▂▂▂▂▂▂▂▂▁▁▁▁▁▁▁▁▁▁▁ ▃
  3.92 μs        Histogram: frequency by time        12.1 μs <

 Memory estimate: 6.88 KiB, allocs estimate: 99.
````

**Comment by avik-pal** [comment]
Benchmarking Script

```julia
import Pkg;
Pkg.activate(@__DIR__);

using BVProblemLibrary, DiffEqDevTools, Plots
using Symbolics, Zygote, Enzyme

using BoundaryValueDiffEq

# Test Case 1
function func_1!(du, u, p, t)
    du[1] = u[2]
    du[2] = 0
end

# Not able to change the initial condition.
# Hard coded solution.
func_1 = ODEFunction(func_1!, analytic = (u0, p, t) -> [5 - t, -1])

function boundary!(residual, u, p, t)
    residual[1] = u[1][1] - 5
    residual[2] = u[length(u)][1]
end

tspan = (0.0, 5.0)
u0 = [5.0, -3.5]

prob = BVProblem(func_1, boundary!, u0, tspan)

alg = MIRK3()

sol = solve(prob, MIRK3(); dt = 0.2)

solve(prob,
    MIRK3(;
        jac_alg = MIRKJacobianComputationAlgorithm(; bc_diffmode = AutoSparseFiniteDiff(),
            collocation_diffmode = AutoSparseFiniteDiff())); dt = 0.2)

jac_algs = [MIRKJacobianComputationAlgorithm(),
    MIRKJacobianComputationAlgorithm(; bc_diffmode = AutoSparseFiniteDiff(),
        collocation_diffmode = AutoSparseFiniteDiff()),
    MIRKJacobianComputationAlgorithm(; bc_diffmode = AutoFiniteDiff(),
        collocation_diffmode = AutoFiniteDiff()),
    MIRKJacobianComputationAlgorithm(; bc_diffmode = AutoSparseEnzyme(),
        collocation_diffmode = AutoSparseFiniteDiff()),
]

alg_names = ["Default", "Auto Detect Sparsity Patterns", "No Sparsity",
    "Auto Sparse Enzyme BC"]

for mirk_method in (MIRK3, MIRK4, MIRK5, MIRK6)
    for (i, jac_alg) in enumerate(jac_algs)
        sol = solve(prob, mirk_method(; jac_alg); dt = 0.2)
        solver = MIRK3(; jac_alg)
        t = @belapsed solve($prob, $solver; dt = 0.2)
        name = alg_names[i]
        println("$(mirk_method) $(name) took $(t)s")
    end
end

for mirk_method in (MIRK3, MIRK4, MIRK5) # , MIRK6)
    for (i, jac_alg) in enumerate(jac_algs)
        sol = solve(prob_bvp_nonlinear_15, mirk_method(; jac_alg); dt = 0.05)
        solver = MIRK3(; jac_alg)
        t = @belapsed solve($prob, $solver; dt = 0.05)
        name = alg_names[i]
        println("$(mirk_method) $(name) took $(t)s")
    end
end
```

**Comment by avik-pal** [comment]
For the TwoPointBVP specialization, we will need some changes into show the problem is specified, holding off on those changes till upstream PRs are ready

**Comment by avik-pal** [comment]
CI needs https://github.com/SciML/OrdinaryDiffEq.jl/pull/2015 https://github.com/SciML/NonlinearSolve.jl/pull/200 https://github.com/SciML/SciMLBase.jl/pull/481

File: src/types.jl [hunk 1]
```diff
@@ -45,4 +45,10 @@ struct BVPSystem{F <: Function, B <: Union{Function, SciMLBase.TwoPointBVPFuncti
     tmp::tType                  # M
 end
 
-Base.eltype(S::BVPSystem) = eltype(S.tmp)
+Base.eltype(S::BVPSystem) = eltype(S.tmp.du)
+
+# Sparsity Detection
+Base.@kwdef struct MIRKJacobianComputationAlgorithm{BD, CD}
+    bc_diffmode::BD = AutoFiniteDiff()
+    collocation_diffmode::CD = AutoSparseFiniteDiff()
```
Comment (ChrisRackauckas) [inline src/types.jl:53]: we normally default to forward diff

Comment (avik-pal) [reply src/types.jl:53]: The forward diff implementation is not entirely efficient yet. I just slapped some `DiffCache`s around to get it to work but it warns that the size of the diffcache is not proper and might hurt performance.

Comment (ChrisRackauckas) [reply src/types.jl:53]: resizing will require the diffcache to resize, but that's fine (expected) in this context when the adaptivity occurs. Though what you'd need to add is a `resize!` on the diffcache when the adaptivity happens in order to make it more direct.

Comment (avik-pal) [reply src/types.jl:53]: Yes but here I am getting a resize warning even when the mesh is not resized which makes me think I am somehow incorrectly using DiffCache

Comment (avik-pal) [reply src/types.jl:53]: I think I figured out the problem. I need to use the chunksize of `y`

**Comment by avik-pal** [comment]
@ChrisRackauckas this is ready from my end

**Review by ChrisRackauckas — APPROVED** [review]

**Pull request merged**